File: src/org/scy/hullmods/SCY_engineering.java
package org.scy.hullmods;

import static org.scy.SCY_settingsData.engineering_noncompatible;
import static org.scy.SCY_txt.txt;
import static org.scy.StarficzAIUtils.DEBUG_ENABLED;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.combat.*;
import com.fs.starfarer.api.combat.ShipAPI.HullSize;
import com.fs.starfarer.api.combat.listeners.AdvanceableListener;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.util.IntervalUtil;
import com.fs.starfarer.api.util.Misc;
import com.fs.starfarer.api.util.Pair;
import com.fs.starfarer.combat.CombatEngine;
import com.fs.starfarer.combat.tasks.CombatTaskManager;
import org.lazywizard.lazylib.ui.FontException;
import org.lazywizard.lazylib.ui.LazyFont;
import org.magiclib.subsystems.MagicSubsystemsManager;
import org.scy.StarficzAIUtils;
import org.lwjgl.util.vector.Vector2f;
import org.magiclib.util.MagicIncompatibleHullmods;
import org.scy.subsystems.EngineJumpstart;

import java.awt.*;
import java.util.*;
import java.util.List;
import java.util.stream.Collectors;

public class SCY_engineering extends BaseHullMod {

  private final Map<HullSize, Float> VENTING_BONUS = new HashMap<>();
  {
    VENTING_BONUS.put(HullSize.FIGHTER, 5f);
    VENTING_BONUS.put(HullSize.FRIGATE, 5f);
    VENTING_BONUS.put(HullSize.DESTROYER, 3f);
    VENTING_BONUS.put(HullSize.CRUISER, 2f);
    VENTING_BONUS.put(HullSize.CAPITAL_SHIP, 1f);
  }

  private final float SENSOT_STILL = -25, SENSOR_MOVE = 25;
  private String ID;
  private float VENT_MULT = 3f;
  private float CAP_MULT = 2f;


  @Override
  public void applyEffectsBeforeShipCreation(HullSize hullSize, MutableShipStatsAPI stats, String id) {
    stats.getVentRateMult().modifyMult(id, VENT_MULT);
    stats.getEngineDamageTakenMult().modifyMult(id, 0.75f);
    stats.getCombatEngineRepairTimeMult().modifyMult(id, 0.75f);
    ID = id;
  }

  @Override
  public void applyEffectsAfterShipCreation(ShipAPI ship, String id) {
    for (String tmp : engineering_noncompatible) {
      if (ship.getVariant().getHullMods().contains(tmp)) {
        MagicIncompatibleHullmods.removeHullmodWithWarning(ship.getVariant(), tmp, "SCY_engineering");
      }
    }
    ship.getMutableStats().getFluxCapacity().modifyMult(id, CAP_MULT);
    ship.getMutableStats().getVentRateMult().modifyPercent(id, ship.getVariant().getNumFluxCapacitors()*VENTING_BONUS.get(ship.getHullSize()));
    if(ship.getHullSize() != HullSize.FIGHTER) {
      MagicSubsystemsManager.addSubsystemToShip(ship, new EngineJumpstart(ship));
      if(!ship.hasListenerOfClass(SCYVentingAI.class)) ship.addListener(new SCYVentingAI(ship));
    }
  }

  public static class SCYVentingAI implements AdvanceableListener{
    public IntervalUtil damageTracker = new IntervalUtil(0.2F, 0.3F);
    public CombatEngineAPI engine;
    public ShipAPI ship;
    public ShipAPI target;
    public Vector2f backupPoint;
    public float lastUpdatedTime = 0f;
    public List<StarficzAIUtils.FutureHit> incomingProjectiles = new ArrayList<>();
    public List<StarficzAIUtils.FutureHit> predictedWeaponHits = new ArrayList<>();
    public List<StarficzAIUtils.FutureHit> combinedHits = new ArrayList<>();
    SCYVentingAI(ShipAPI ship){
      this.ship = ship;
    }
    public LazyFont font;
    {
        try {
            font = LazyFont.loadFont("graphics/fonts/insignia15LTaa.fnt");
        } catch (FontException e) {
            throw new RuntimeException(e);
        }
    }
    public LazyFont.DrawableString incomingProjectilesString = font.createText("", Color.orange, 15f);
    public LazyFont.DrawableString predictedWeaponHitsString = font.createText("", Color.orange, 15f);

    @Override
    public void advance(float amount) {
      engine = Global.getCombatEngine();
      if (!ship.isAlive() || ship.getParentStation() != null || engine == null || !engine.isEntityInPlay(ship)) {
        return;
      }

      // Calculate Decision Flags
      float bufferTime = ship.getShield() != null ? ship.getShield().getUnfoldTime()/(ship.getShield().getArc()/90f) : 0f + 0.2f; // 0.2 sec of buffer time before getting hit

      damageTracker.advance(amount);
      if (damageTracker.intervalElapsed()) {
        lastUpdatedTime = Global.getCombatEngine().getTotalElapsedTime(false);
        incomingProjectiles = StarficzAIUtils.incomingProjectileHits(ship, ship.getLocation());
        float timeToPredict = Math.max(ship.getFluxTracker().getTimeToVent() + bufferTime + damageTracker.getMaxInterval(), 3f);
        predictedWeaponHits = StarficzAIUtils.generatePredictedWeaponHits(ship, ship.getLocation(), timeToPredict);
        combinedHits.clear();
        combinedHits.addAll(incomingProjectiles);
        combinedHits.addAll(predictedWeaponHits);
        if(ship.getFluxLevel() > 0.7f) backupPoint = StarficzAIUtils.getBackingOffStrafePoint(ship);
        if(DEBUG_ENABLED){
          incomingProjectilesString.setText(incomingProjectiles.stream()
                  .map(StarficzAIUtils.FutureHit::toString)
                  .collect(Collectors.joining(System.lineSeparator())));
          predictedWeaponHitsString.setText(predictedWeaponHits.stream()
                  .map(StarficzAIUtils.FutureHit::toString)
                  .collect(Collectors.joining(System.lineSeparator())));
        }
      }

      // calculate how much damage the ship would take if unphased/vent/used system
      float currentTime = Global.getCombatEngine().getTotalElapsedTime(false);
      float timeElapsed = currentTime - lastUpdatedTime;

      float armorBase = StarficzAIUtils.getCurrentArmorRating(ship);
      float armorMax = ship.getArmorGrid().getArmorRating();
      float armorMinLevel = ship.getMutableStats().getMinArmorFraction().getModifiedValue();
      float armorVent = armorBase;

      float hullDamageIfVent = 0f;
      float empDamageIfVent = 0f;

      float mountHP = 0f;
      for (WeaponAPI weapon : ship.getAllWeapons()) {
        mountHP += weapon.getCurrHealth();
      }
      float empDamageLevelVent = empDamageIfVent / mountHP;

      for (StarficzAIUtils.FutureHit hit : combinedHits) {
        float timeToHit = (hit.timeToHit - timeElapsed);
        if (timeToHit < -0.1f) continue; // skip hits that have already happened
        if (timeToHit < ship.getFluxTracker().getTimeToVent() + bufferTime) {
          Pair<Float, Float> trueDamage = StarficzAIUtils.damageAfterArmor(hit.damageType, hit.damage, hit.hitStrength, armorVent, ship);
          hullDamageIfVent += trueDamage.two;
          empDamageIfVent += hit.empDamage;
          armorVent = Math.max(armorVent - trueDamage.one, armorMinLevel * armorMax);
        }
      }

      float armorDamageLevelVent = (armorBase - armorVent) / armorMax;
      float hullDamageLevelVent = hullDamageIfVent / (ship.getHitpoints() * ship.getHullLevel());

      ShipVariantAPI variant = Global.getSettings().getVariant(ship.getHullSpec().getBaseHullId() + "_combat");
      float modules = variant == null ? 0 : variant.getStationModules().size();

      float alive = 0;
      for(ShipAPI module : ship.getChildModulesCopy()) {
        if (module.getHitpoints() <= 0f) continue;
        alive++;
      }

      float damageRiskMult = Misc.interpolate(1,5, modules > 0 ? alive/modules : 0);


      if (!engine.isUIAutopilotOn() || engine.getPlayerShip() != ship) {
        if(armorDamageLevelVent < 0.01f && hullDamageLevelVent < 0.01f && empDamageLevelVent < 0.5f){
          ship.getAIFlags().setFlag(ShipwideAIFlags.AIFlags.DO_NOT_BACK_OFF, 0.05f);
          ship.getAIFlags().setFlag(ShipwideAIFlags.AIFlags.DO_NOT_BACK_OFF_EVEN_WHILE_VENTING, 0.05f);
        }

        // vent control
        if (ship.getFluxLevel() > 0.2f && armorDamageLevelVent < (0.03f*damageRiskMult) && hullDamageLevelVent < (0.03f*damageRiskMult) && empDamageLevelVent < (0.5f*damageRiskMult)) {
          ship.giveCommand(ShipCommand.VENT_FLUX, null, 0);
        } else {
          ship.blockCommandForOneFrame(ShipCommand.VENT_FLUX);
          if(ship.getFluxLevel() > 0.8f && backupPoint != null && !ship.getFluxTracker().isVenting()){
            CombatTaskManager taskManager = ((CombatEngine) engine).getTaskManager(this.ship.getOwner(), this.ship.isAlly());
            if (taskManager == null || taskManager.getAssignmentFor(ship) == null || taskManager.getAssignmentFor(ship).getType() != CombatAssignmentType.INTERCEPT) {
              StarficzAIUtils.strafeToPointV2(ship, backupPoint);
            }
          }
        }
      }
    }
  }

  @Override
  public String getDescriptionParam(int index, ShipAPI.HullSize hullSize) {
    if (index == 0) return "25" + txt("%");
    if (index == 1) return CAP_MULT + "x";
    if (index == 2) return VENT_MULT + "x";
    if (index == 3) return "+5/3/2/1" + txt("%");

    return null;
  }

  @Override
  public void advanceInCampaign(FleetMemberAPI member, float amount) {
    if (member.getFleetData() != null
        && member.getFleetData().getFleet() != null
        && !member.isMothballed()
        && !member.getFleetCommander().isPlayer()) {
      if (member.getFleetData().getFleet().getCurrBurnLevel() < 3) {
        //                member.getStats().getSensorProfile().modifyPercent(ID, -SENSOR_OFFSET);
        member.getStats().getSensorProfile().modifyPercent(ID, SENSOT_STILL);
      } else {
        //                member.getStats().getSensorProfile().modifyPercent(ID, +SENSOR_OFFSET);
        member.getStats().getSensorProfile().modifyPercent(ID, SENSOR_MOVE);
      }
    }
  }
}

File: src/org/scy/hullmods/ScyEngineering.kt
package org.scy.hullmods

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.combat.*
import com.fs.starfarer.api.combat.ShipAPI.HullSize
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.ui.Alignment
import com.fs.starfarer.api.ui.TooltipMakerAPI
import com.fs.starfarer.api.util.Misc
import org.magiclib.kotlin.getGoSlowBurnLevel
import org.magiclib.subsystems.MagicSubsystemsManager.addSubsystemToShip
import org.magiclib.util.MagicIncompatibleHullmods
import org.scy.*
import org.scy.combatai.ScyAiV2
import org.scy.subsystems.EngineJumpstart

class ScyEngineering: BaseHullMod() {
    private val VENT_MULT = 3f
    private val CAP_MULT = 2f
    private val SLOW_SUPPLIES_PERCENT = -50f
    private val SLOW_PROFILE_PERCENT = -25f
    private val BURN_PROFILE_PERCENT = 25f
    private val ENGINE_HEALTH_PERCENT = 25f
    private val VENTING_BONUS = hashMapOf(
        HullSize.FIGHTER to 5f,
        HullSize.FRIGATE to 5f,
        HullSize.DESTROYER to 3f,
        HullSize.CRUISER to 2f,
        HullSize.CAPITAL_SHIP to 1f
    )

    override fun getDisplaySortOrder(): Int {
        return 0
    }

    override fun applyEffectsBeforeShipCreation(hullSize: HullSize?, stats: MutableShipStatsAPI?, id: String?) {
        if(stats == null) return
        stats.ventRateMult.modifyMult(id, VENT_MULT)
        if (hullSize != HullSize.FIGHTER) stats.fluxCapacity.modifyMult(id, CAP_MULT)
        stats.engineHealthBonus.modifyPercent(id, ENGINE_HEALTH_PERCENT)
        stats.combatEngineRepairTimeMult.modifyPercent(id, -ENGINE_HEALTH_PERCENT)
    }

    override fun applyEffectsAfterShipCreation(ship: ShipAPI?, id: String?) {
        if(ship == null) return
        for (tmp in SCY_settingsData.engineering_noncompatible) {
            if (ship.variant.hullMods.contains(tmp)) {
                MagicIncompatibleHullmods.removeHullmodWithWarning(ship.variant, tmp, "SCY_engineering")
            }
        }

        ship.mutableStats.ventRateMult.modifyPercent(id, ship.variant.numFluxCapacitors * VENTING_BONUS[ship.hullSize]!!)
    }

    override fun applyEffectsAfterShipAddedToCombatEngine(ship: ShipAPI, id: String?) {
        // engine jumpstart and custom SCY ai
        if (ship.hullSize != HullSize.FIGHTER && ship.parentStation == null) {
            addSubsystemToShip(ship, EngineJumpstart(ship))
            if (!ship.hasListenerOfClass(ScyAiV2::class.java)) ship.addListener(ScyAiV2(ship))
        }
    }

    override fun advanceInCampaign(member: FleetMemberAPI?, amount: Float) {
        if(member == null) return
        if(member.fleetData?.fleet?.let{ it.currBurnLevel <= it.getGoSlowBurnLevel() } == true){
            member.stats.sensorProfile.modifyPercent("scy_engineering", SLOW_PROFILE_PERCENT)
            member.stats.suppliesPerMonth.modifyPercent("scy_engineering", SLOW_SUPPLIES_PERCENT)
        } else{
            member.stats.sensorProfile.modifyPercent("scy_engineering", BURN_PROFILE_PERCENT)
            member.stats.suppliesPerMonth.unmodify("scy_engineering")
        }
    }

    override fun getDescriptionParam(index: Int, hullSize: HullSize?): String? {
        if (index == 0) return "25" + SCY_txt.txt("%")
        if (index == 1) return CAP_MULT.toString() + "x"
        if (index == 2) return VENT_MULT.toString() + "x"
        if (index == 3) return "+5/3/2/1" + SCY_txt.txt("%")

        return null
    }

    override fun getTooltipWidth(): Float {
        return 410f
    }

    override fun addPostDescriptionSection(tooltip: TooltipMakerAPI, hullSize: HullSize, ship: ShipAPI?, width: Float, isForModSpec: Boolean) {

        val imageHeight = 64f
        val headingPad = 20f
        val underHeadingPad = 10f
        val listPad = 3f

        val activeTextColor = Misc.getTextColor()
        val activeNegativeColor = Misc.getNegativeHighlightColor()
        val activeHeaderBannerColor = Misc.getDarkPlayerColor()
        val activeHeaderTextColor = brighter(Misc.getButtonTextColor(), 0.8f)
        val activeHighlightColor = Misc.getHighlightColor()

        // Scy Engines
        tooltip.addSectionHeading("Scyan Engines", activeHeaderTextColor, activeHeaderBannerColor , Alignment.MID, headingPad)
        tooltip.beginImageWithText(Global.getSettings().getSpriteName("hullmodHeaders", "SCY_engines"), imageHeight*2)
        .apply {
            setBulletedListMode("•")
            setBulletWidth(15f)
            addPara("Increases engine durability by ${ENGINE_HEALTH_PERCENT.toInt()}%%.",
                listPad, activeTextColor, activeHighlightColor, "${ENGINE_HEALTH_PERCENT.toInt()}%")
            addPara("Cuts engine repair time by ${ENGINE_HEALTH_PERCENT.toInt()}%%.",
                listPad, activeTextColor, activeHighlightColor, "${ENGINE_HEALTH_PERCENT.toInt()}%")
            addPara("Engine Jumpstart subsystem instantly reignites flamed out engines. (20 second cooldown)",
                listPad, activeTextColor, activeHighlightColor, "Engine Jumpstart")
            addPara("Incompatible with further engine modifications.",
                activeNegativeColor, listPad)
            addPara("When moving slowly:", listPad)
            setBulletedListMode("    -")
            setBulletWidth(25f)
            addPara("Reduces sensor profile by ${-SLOW_PROFILE_PERCENT.toInt()}%%.",
                1f, activeTextColor, activeHighlightColor, "${-SLOW_PROFILE_PERCENT.toInt()}%")
            addPara("Cuts maintenance costs (supplies/mo) by 50%%.",
                1f, activeTextColor, activeHighlightColor, "${-SLOW_SUPPLIES_PERCENT.toInt()}%")
            setBulletedListMode("•")
            setBulletWidth(15f)
            addPara("Otherwise:", listPad)
            setBulletedListMode("    -")
            setBulletWidth(25f)
            addPara("Increases sensor profile by ${BURN_PROFILE_PERCENT.toInt()}%%.",
                1f, activeTextColor, activeNegativeColor, "${BURN_PROFILE_PERCENT.toInt()}%")
        }
        tooltip.addImageWithText(underHeadingPad)

        // Scy Flux
        tooltip.addSectionHeading("Scyan Flux Grid", activeHeaderTextColor, activeHeaderBannerColor , Alignment.MID, headingPad)
        tooltip.beginImageWithText(Global.getSettings().getSpriteName("hullmodHeaders", "SCY_flux"), imageHeight)
        .apply {
            setBulletedListMode("•")
            setBulletWidth(15f)
            addPara("Increases flux capacity by x${CAP_MULT.toInt()} from all sources.",
                listPad, activeTextColor, activeHighlightColor, "x${CAP_MULT.toInt()}")
            addPara("Increases flux dissipation rate while actively venting by x${VENT_MULT.toInt()}. (x6 base dissipation rate)",
                listPad, activeTextColor, activeHighlightColor, "x${VENT_MULT.toInt()}", "x2", "x6")
            addPara("Further increases flux dissipation rate while actively venting by " +
                    "${VENTING_BONUS[HullSize.FRIGATE]!!.toInt()}%%/" +
                    "${VENTING_BONUS[HullSize.DESTROYER]!!.toInt()}%%/" +
                    "${VENTING_BONUS[HullSize.CRUISER]!!.toInt()}%%/" +
                    "${VENTING_BONUS[HullSize.CAPITAL_SHIP]!!.toInt()}%% per flux capacitor.",
                listPad, activeTextColor, activeHighlightColor,
                "${VENTING_BONUS[HullSize.FRIGATE]!!.toInt()}%",
                "${VENTING_BONUS[HullSize.DESTROYER]!!.toInt()}%",
                "${VENTING_BONUS[HullSize.CRUISER]!!.toInt()}%",
                "${VENTING_BONUS[HullSize.CAPITAL_SHIP]!!.toInt()}%")
        }
        tooltip.addImageWithText(underHeadingPad)
    }
}